web/packages/teleterm/src/ui/DocumentCluster/useUserPreferences.ts

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright 2023 Gravitational, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useCallback, useEffect, useRef } from 'react';
+
+import {
+  useAsync,
+  makeSuccessAttempt,
+  mapAttempt,
+  Attempt,
+} from 'shared/hooks/useAsync';
+
+import {
+  DefaultTab,
+  ViewMode,
+  UnifiedResourcePreferences,
+} from 'shared/services/unifiedResourcePreferences';
+
+import { useAppContext } from 'teleterm/ui/appContextProvider';
+
+import { routing, ClusterUri } from 'teleterm/ui/uri';
+
+import { UserPreferences } from 'teleterm/services/tshd/types';
+
+export function useUserPreferences(clusterUri: ClusterUri): {
+  userPreferencesAttempt: Attempt<UserPreferences>;
+  unifiedResourcePreferencesFallback: UnifiedResourcePreferences;
+  updateUserPreferencesAttempt: Attempt<void>;
+  updateUserPreferences(newPreferences: UserPreferences): Promise<void>;
+} {
+  const appContext = useAppContext();
+  const [
+    fetchUserPreferencesAttempt,
+    runFetchUserPreferencesAttempt,
+    setFetchUserPreferencesAttempt,
+  ] = useAsync<Promise<UserPreferences>[], UserPreferences>(value => value);
+  const [updateUserPreferencesAttempt, runUpdateUserPreferencesAttempt] =
+    useAsync(async (newPreferences: UserPreferences) =>
+      appContext.tshd.updateUserPreferences({
+        clusterUri,
+        userPreferences: newPreferences,
+      })
+    );
+  appContext.workspacesService.useState();
```

Comment:
This has some heavy implications when it comes to performance, do we absolutely need it here? I feel like this should be reserved only to components which are not bound to a particular workspace.

Reply:
Yeah, I think we can remove it.